crates/external_services/src/grpc_client/dynamic_routing.rs

```diff
@@ -0,0 +1,213 @@
+)]
+pub mod success_rate {
+    tonic::include_proto!("success_rate");
+}
+
+pub type DRResult<T> = CustomResult<T, DRError>;
+
+#[derive(Debug, Clone, thiserror::Error)]
+pub enum DRError {
+    #[error("Error buling the gRPC Client for communication")]
+    ClientBuildingFailed,
+    #[error("Error getting a response from the gRPC Server")]
+    GrpcServerResponseFailure,
+}
+
+// Struct consists of all the services provided by the client
+#[derive(Debug, Clone)]
+pub struct RoutingStrategy {
+    pub success_rate_client: Option<SuccessRateCalculatorClient<Channel>>,
+}
+
+#[derive(Debug, Clone, serde::Deserialize, serde::Serialize, Default)]
+pub struct DynamicRoutingClientConfig {
+    pub host: String,
+    pub port: u16,
+    pub enabled: bool,
+}
+
+// establish connection with the server
+impl DynamicRoutingClientConfig {
+    pub async fn get_dynamic_routing_connection(
+        self,
+    ) -> Result<RoutingStrategy, Box<dyn std::error::Error>> {
+        let success_rate_client = if self.enabled {
+            let uri = format!("http://{}:{}", self.host, self.port);
+            let channel = tonic::transport::Endpoint::new(uri)?.connect().await?;
+            Some(SuccessRateCalculatorClient::new(channel))
+        } else {
+            None
+        };
+        Ok(RoutingStrategy {
+            success_rate_client,
+        })
+    }
+}
+
+// The trait Success Based Dynamic Routing would have the functions required to support the calculation and updation window
+#[async_trait::async_trait]
+pub trait SuccessBasedDynamicRouting: dyn_clone::DynClone + Send + Sync {
+    async fn calculate_success_rate(
+        &self,
+        id: id_type::ProfileId,
+        dynamic_routing_config: SuccessBasedRoutingConfig,
+        label_input: Vec<RoutableConnectorChoice>,
+    ) -> DRResult<CalSuccessRateResponse>;
+
+    async fn update_success_rate(
+        &self,
+        id: id_type::ProfileId,
+        dynamic_routing_config: SuccessBasedRoutingConfig,
+        response: Vec<RoutableConnectorChoiceWithStatus>,
+    ) -> DRResult<UpdateSuccessRateWindowResponse>;
+}
+
+#[async_trait::async_trait]
+impl SuccessBasedDynamicRouting for SuccessRateCalculatorClient<Channel> {
+    async fn calculate_success_rate(
+        &self,
+        id: id_type::ProfileId,
+        dynamic_routing_config: SuccessBasedRoutingConfig,
+        label_input: Vec<RoutableConnectorChoice>,
+    ) -> DRResult<CalSuccessRateResponse> {
+        let params = dynamic_routing_config
+            .params
+            .map(|vec| {
+                if !vec.is_empty() {
+                    vec.into_iter()
+                        .map(|param| param.to_string())
+                        .collect::<Vec<_>>()
+                        .join(":")
+                } else {
+                    String::default()
+                }
+            })
+            .get_required_value("Vector of params")
+            .change_context(DRError::ClientBuildingFailed)?;
```

Comment:
Why is this mapped to client build failure?

Reply:
Cause the params needed to build the client was empty.

Reply:
client is already built during startup right. This can happen when the params itself is not found in db for a profile as it is Option. I think we should say necessary params not found for making request

Reply:
Ok